src/azul/plugins/metadata/hca/transform.py

```diff
@@ -840,39 +876,45 @@ def _sequencing_input(self, sequencing_input: api.Biomaterial) -> MutableJSON:
     @classmethod
     def _sample_types(cls) -> FieldTypes:
         return {
+            'document_id': null_str,
+            'biomaterial_id': null_str,
             'entity_type': null_str,
+            'organ': null_str,
+            'organ_part': null_str,
+            'model_organ': null_str,
+            'model_organ_part': null_str,
             'effective_organ': null_str,
-            **cls._cell_line_types(),
-            **cls._organoid_types(),
-            **cls._specimen_types()
         }
 
-    def _sample(self, sample: api.Biomaterial) -> MutableJSON:
-        # Start construction of a `sample` inner entity by including all fields
-        # possible from any entities that can be a sample. This is done to
-        # have consistency of fields between various sample inner entities
-        # to allow Elasticsearch to search and sort against these entities.
-        sample_ = dict.fromkeys(ChainMap(
-            self._cell_line_types(),
-            self._organoid_types(),
-            self._specimen_types()
-        ).keys())
-        entity_type, entity_dict = (
-            'cell_lines', self._cell_line(sample)
-        ) if isinstance(sample, api.CellLine) else (
-            'organoids', self._organoid(sample)
-        ) if isinstance(sample, api.Organoid) else (
-            'specimens', self._specimen(sample)
-        ) if isinstance(sample, api.SpecimenFromOrganism) else (
-            require(False, sample), None
-        )
-        sample_.update(entity_dict)
-        sample_['entity_type'] = entity_type
-        assert hasattr(sample, 'organ') != hasattr(sample, 'model_organ')
-        sample_['effective_organ'] = sample.organ if hasattr(sample, 'organ') else sample.model_organ
-        assert sample_['document_id'] == str(sample.document_id)
-        assert sample_['biomaterial_id'] == sample.biomaterial_id
-        return sample_
+    def _samples(self, samples: Iterable[api.Biomaterial]) -> JSON:
+        """
+        Returns inner entity values with all given samples represented both in a
+        'samples' inner entity and a type-specific 'sample_{entity_type}'.
+        The 'samples' inner entity is a minified polymorphic field containing
+        the few properties common to all samples. This allows facets on these
+        properties regardless of the sample entity type.
+        """
+
+        inner_entities = defaultdict(list)
+        inner_entities['samples'] = []
+        entity_types = [
+            ('cell_lines', api.CellLine, self._cell_line, self._sample_cell_line),
+            ('organoids', api.Organoid, self._organoid, self._sample_organoid),
+            ('specimens', api.SpecimenFromOrganism, self._specimen, self._sample_specimen),
+        ]
+        for entity_type, api_class, entity_fn, sample_fn in entity_types:
+            entities = [entity_fn(s) for s in samples if isinstance(s, api_class)]
```

Comment:
Consider eliminating the indirection (API object -> foo JSON -> sample_foo JSON) and doing (API object -> sample_foo JSON). That would eliminate the ambiguity about where to place the methods. They would all follow `_sample_types`.